cli/azd/pkg/azsdk/zip_deploy_client.go

```diff
@@ -114,6 +122,202 @@ func (c *ZipDeployClient) BeginDeploy(
 	return runtime.NewPoller(response, c.pipeline, pollerOptions)
 }
 
+// Deploys the specified application zip to the azure app service using deployment status api and waits for completion
+func (c *ZipDeployClient) BeginDeployTrackStatus(
+	ctx context.Context,
+	zipFile io.ReadSeeker,
+	subscriptionId,
+	resourceGroup,
+	appName string,
+) (*runtime.Poller[armappservice.WebAppsClientGetProductionSiteDeploymentStatusResponse], error) {
+	request, err := c.createDeployRequest(ctx, zipFile)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := c.pipeline.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	if !runtime.HasStatusCode(response, http.StatusAccepted) {
+		return nil, runtime.NewResponseError(response)
+	}
+
+	client, err := armappservice.NewWebAppsClient(subscriptionId, c.cred, nil)
+
+	if err != nil {
+		return nil, fmt.Errorf("creating web app client: %w", err)
+	}
+
+	deploymentStatusId := response.Header.Get("Scm-Deployment-Id")
+	if deploymentStatusId == "" {
+		return nil, fmt.Errorf("empty deployment status id")
+	}
+
+	// nolint:lll
+	// Example definition: https://github.com/Azure/azure-rest-api-specs/tree/main/specification/web/resource-manager/Microsoft.Web/stable/2022-03-01/examples/GetSiteDeploymentStatus.json
+	var poller *runtime.Poller[armappservice.WebAppsClientGetProductionSiteDeploymentStatusResponse]
+	err = retry.Do(
+		ctx,
+		retry.WithMaxRetries(3, retry.NewConstant(5*time.Second)),
+		func(ctx context.Context) error {
+			pollerRetry, err := client.BeginGetProductionSiteDeploymentStatus(ctx, resourceGroup, appName, deploymentStatusId, nil)
+			if err != nil {
+				var httpErr *azcore.ResponseError
+				if errors.As(err, &httpErr) {
+					// 404 - Retry if the resource is not found as deployment id doesn't match with temp deployment id
```

Comment:
"// 404 - Retry if the resource is not found as deployment id doesn't match with temp deployment id"

Without any context, I'm not getting the full details of why we would retry for 404s out of this comment. I'm not sure if this matches your expectations, but I think the reason would be `// We get temporary 404s when the ARM API hasn't fully committed the deployment record with the given deployment ID`.